lib/pure/collections/sets.nim

```diff
@@ -907,7 +909,52 @@ iterator pairs*[A](s: OrderedSet[A]): tuple[a: int, b: A] =
   forAllOrderedPairs:
     yield (idx, s.data[h].key)
 
-
+proc fromJsonHook*[A, B](s: var SomeSet[A], jsonNode: B) =
+  ## Enables `fromJson` for `HashSet` and `OrderedSet` types.
+  ## 
+  ## See also:
+  ## * `toJsonHook proc<#toJsonHook,SomeSet[A]>`_
+  runnableExamples:
+    import std/[json, jsonutils]
+    type
+      Foo = object
+        hs: HashSet[string]
+        os: OrderedSet[string]
+    var foo: Foo
+    fromJson(foo, parseJson("""
+      {"hs": ["hash", "set"], "os": ["ordered", "set"]}"""))
+    assert foo.hs == ["hash", "set"].toHashSet
+    assert foo.os == ["ordered", "set"].toOrderedSet
+
+  mixin jsonTo
+  assert jsonNode.kind == JArray,
+          "The kind of the `jsonNode` must be JArray, but its actual " &
+          "type is " & $jsonNode.kind & "."
+  clear(s)
+  for v in jsonNode:
+    incl(s, jsonTo(v, A))
+
+proc toJsonHook*[A](s: SomeSet[A]): auto =
+  ## Enables `toJson` for `HashSet` and `OrderedSet` types.
+  ##
+  ## See also:
+  ## * `fromJsonHook proc<#fromJsonHook,SomeSet[A],B>`_
+  runnableExamples:
+    import std/[json, jsonutils]
+    type
+      Foo = object
+        hs: HashSet[string]
+        os: OrderedSet[string]
+    let foo = Foo(
+      hs: ["hash"].toHashSet,
+      os: ["ordered", "set"].toOrderedSet)
+    assert $toJson(foo) == """{"hs":["hash"],"os":["ordered","set"]}"""
```

Comment:
optional nit: you can make this test shorter (good for a runnableExamples) by using a tuple containing `hs`, `os` instead of introducing `Foo`
(likewise in the other runnableExample IIRC)

Reply:
Fixed.